Overhaul the `BASICS` part of the documentation

alalazo opened the pull request:
The intent is to organize the information and sections we have in a better way, remove outdated content, and add what is missing.

Modifications:
- [x] Added an `INTRODUCTION` part, including
  1. The "Feature Overview"
  1. A very short "Getting Started" guide
  1. An explanation of the Spec syntax
  1. Installing Spack prerequisites on `linux` and `macOS`
  1. Using Spack on Windows
- [x] Added a `BASIC USAGE` part, including
  1. Listing, installing, uninstalling, finding and using packages
  2. Configuring compilers
  3. Using Spack environments with unify:true and a view
- [x] Added an `ADVANCED TOPICS` part, including
  1. Toolchains
  2. Auditing configuration and packages
  3. Verifying installations

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
Reworked "Verify your installation of Spack"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add "List packages you can install"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Limit max height of code blocks

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Added "Installing your first package"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move a few sections to "Basic usage"

Deleted "Clean environment", since it doesn't make sense
anymore, as it is written.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "Configure compilers" to "Basic usage"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "Vendor Specific ..." to "Basic usage"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add "Use the software you just installed"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "Toolchains"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove tip telling users to run --insecure + warning

It shouldn't be part of the getting started guide

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove the "System packages" section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove the "Package utilities" section

Should be covered by "Verify your installation"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "GPG signing" to Basic Usage

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "Spack On Windows" into its own top-level section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Create an "Advanced Topics" section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Reworked "Configuring Compilers"

Removed redundant, or outdated sections.
Added examples of compilers listed from a buildcache.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove "Vendor specific ...", as it is outdated

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Moved "Fortran compilers on macOS" in a different section

This may need some rethinking...

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move RPATH specification for compilers to the correct section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Moved a few topics to the Advanced section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove "Shell Support" and "GPG Signing"

Both sections are redundant, and outdated

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Recover GPG key creation and trusting/untrusting

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove "Bootstrapping clingo"

Redundant, since this is covered in the Reference Bootstrapping

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Extract a Spec Syntax section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Restructure the index

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove outdated parts of Spec Syntax

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Reorder existing topics

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove outdated content from "Spack Environment" section

Co-authored-by: Alec Scott <19558067+alecbcs@users.noreply.github.com>
Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "Spack Prerequisites" at the end of the introduction

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Move "Spack On Windows" in the INTRODUCTION

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/docs/_static/css/custom.css:
```diff
@@ -0,0 +1,3 @@
+.rst-content div[class^="highlight"] pre {
+    max-height: 300px;
+}
```
Not a fan to apply this globally (also `.highlight` is the sensible selector)

alalazo committed:
Tweak details in "Getting Started" section

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,712 @@
+built using ``gcc`` at version 4.7.5 for a generic ``x86_64`` architecture,
+with debug options enabled, and without Qt support. Additionally, it
+says to link it with the ``callpath`` library (which it depends on),
+and to build callpath with ``gcc`` 4.7.2. Most specs will not be as
```
The spec does not say "to link it with", it says it has to depend on it.

Maybe also update the compiler example to something more sensible, like `%gcc@13`.

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,712 @@
+  ``cflags``, ``cxxflags``, ``fflags``, ``cppflags``, ``ldflags``, and ``ldlibs``.
+  Use ``name==<value>`` to propagate compiler flags through the dependencies.
+* ``target=<value> os=<value>`` Optional architecture specifier
+  (e.g., ``target=haswell os=CNL10``)
```
platform is missing, os is an unfamiliar example to all readers.

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,712 @@
+There are two things to notice here. The first is that specs are
+recursively defined. That is, each dependency after ``%`` or ``^`` is
+a spec itself. The second is that everything is optional *except* for
+the initial package name identifier. Users can be as vague or as
```
not the case, depends on context...

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,712 @@
+DAG structure. Each ``^`` in the full spec refers to some dependency
+of the root package. Spack will raise an error if you supply a name
+after ``^`` that the root does not actually depend on (e.g., ``mpileaks
+^emacs@23.3``). Each ``%`` refers to some direct dependency, and Spack will
+similarly raise an error if that relationship is invalid.
```
why so much focus on "raising an error" :thinking: it's about the meaning of ^ and %?

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,712 @@
+similarly raise an error if that relationship is invalid.
+
+Spack further simplifies things by only allowing one configuration of
+each package within the link/run + direct build dependencies of a
```
dependency types were not introduced. If the intention is to be vague, just talk about runtime dependencies.

alalazo replied:
So far, I just moved the content up for `Spec Syntax`. I was planning to improve on it in a following PR, and add diagrams etc.

haampie replied:
I would just make all changes in one PR. Having URLs change constantly is bad for search engines and sharing links to docs/sections.

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,712 @@
+in the range ``1.0:1.5`` and the specific version ``1.7.1``.
+
+-----------------
+Binary Provenance
```
Feels out of place. It's highly specific, and the generic variants section is below it.

alalazo replied:
Moved down in the `Variant` section

alalazo committed:
Improve "Spec Syntax"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Revert "Limit max height of code blocks"

This reverts commit 48a4df083eee347a3cf521fc2ec1d5c9de88236c.

tgamblin commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,263 @@
+Defining and Using Toolchains
+=============================
+
+Spack can be configured to associate certain combinations of specs for
+easy reference on the command line and in config and environment
+files. These combinations are called ``toolchains``, because their
```
Can we start with a motivating example instead of with the definition? e.g.:

> Spack lets you specify compilers on the CLI with, e.g., `%gcc` or `%c,cxx=clang %fortran=gcc`, and you can specify flags with `cflags`, `cxxflags`, and `fflags`. Depending on how complex your compiler setup is, it can be cumbersome to specify all of your preferences on the CLI. Spack has a special type configuration called  *toolchains*, which let you encapsulate the configuration for compilers, other libraries, and flags into a single name that you can reference as though it were one option.

tgamblin commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,263 @@
+files. These combinations are called ``toolchains``, because their
+primary intended use is for associating compiler combinations to
+apply. Toolchains are referenced by name like a direct dependency,
+using the ``%`` sigil. There are two styles of toolchain config, one
+using conditional dependencies through the spec syntax and one with
+conditionals explicitly in the yaml:
+
+.. code-block:: yaml
+
+   toolchains:
+     gcc_all: cflags=-O3 '%[when=%c virtuals=c]gcc %[when=%cxx virtuals=cxx]gcc %[when=%fortran virtuals=fortran]gcc'
+     llvm_gfortran:
+     - spec: cflags=-O3
+     - spec: '%[virtuals=c]llvm'
+       when: '%c'
+     - spec: '%[virtuals=cxx]llvm'
+       when: '%cxx'
+     - spec: '%[virtuals=fortran]gcc'
+       when: '%fortran'
```
why document both styles?  I would start with the simple style and the `%c=llvm` syntax, and then document the longer syntax later (if at all)... I don't think it's interesting to users that there are two ways to do this.

alalazo replied:
I'll improve this, but note that toolchains are in exact same state as was approved in #50481 3 weeks ago. :grimacing:

tgamblin commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,263 @@
+     - spec: '%[virtuals=fortran]gcc'
+       when: '%fortran'
+
+The two syntaxes are equivalent. It is not necessary to use
+conditional dependencies with toolchains, but in most cases it his
+highly recommended. Similarly, while any spec constraint can be
```
I think it would help to show a matrix to talk about why it's recommended.

tgamblin reviewed (changes requested):
I didn't get very far but I have some requests already

alalazo committed:
Address review comments for toolchains

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Split "basic_usage" into two files

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove documentation on `spack deprecate`

The documentation takes more space on warning the
user that the command should not be used in most
cases, than documenting what it does.

The use case should be completely covered by splicing,
so remove docs for now to avoid advertizing it.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+
+The package name identifier in the spec is the root of some dependency DAG, and the DAG itself is implicit.
+Spack knows the precise dependencies among packages, but users do not need to know the full DAG structure.
+Each ``^`` in the full spec refers to some *transitive* dependency of the root package.
```
```suggestion
Each ``^`` in the full spec refers to a *transitive* dependency of the root package.
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+The package name identifier in the spec is the root of some dependency DAG, and the DAG itself is implicit.
+Spack knows the precise dependencies among packages, but users do not need to know the full DAG structure.
+Each ``^`` in the full spec refers to some *transitive* dependency of the root package.
+Each ``%`` refers to some *direct* dependency, either of the root, or of the last defined transitive dependency .
```
```suggestion
Each ``%`` refers to a *direct* dependency, either of the root, or of the last defined transitive dependency .
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+Each ``^`` in the full spec refers to some *transitive* dependency of the root package.
+Each ``%`` refers to some *direct* dependency, either of the root, or of the last defined transitive dependency .
+
+Spack takes care of allowing only a single configuration of each package, where that is needed for consistency.
```
"takes care of allowing" sounds weird

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+You cannot build an ``mpileaks`` version that depends on one version of ``mpich`` *and* on a ``callpath`` version that depends on some *other* version of ``mpich``.
+In general, such a configuration would likely behave unexpectedly at runtime, and Spack enforces this to ensure a consistent runtime environment.
+
+The point of specs is to abstract this full DAG from Spack users.
```
```suggestion
The purpose of specs is to abstract this full DAG away from Spack users.
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+Variants are named options associated with a particular package and are typically used to enable or disable certain features at build time.
+They are optional, as each package must provide default values for each variant it makes available.
+
+The names of variants available for a particular package depend on what was provided by the package author.
```
```suggestion
The variants available for a particular package are defined by the package author.
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+notable exceptions such as git branches and tags i.e ``pkg@develop``.
+These versions do not naturally have source provenance because they refer to a range
+of commits (branches) or can be changed outside the spack packaging infrastructure
+(tags). Without source provenace we can not have binary provenance.
```
```suggestion
(tags). Without source provenance we cannot have binary provenance.
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+by a space. The command ``spack install libelf cppflags="-O3"%intel``
+will be interpreted as an attempt to set ``cppflags="-O3%intel"``.
+
+The six compiler flags are injected in the order of implicit make commands
```
```suggestion
The six compiler flags are injected in the same order as implicit make commands
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+   $ spack install zlib%gcc@5.5.0 target=icelake
+   ==> Error: cannot produce optimized binary for micro-architecture "icelake" with gcc@5.5.0 [supported compiler versions are 8:]
+
+When instead an old compiler is selected on a recent enough microarchitecture but there is
+no explicit ``target`` specification, Spack will optimize for the best match it can find instead
```
```suggestion
Conversely, if an old compiler is selected for a newer microarchitecture, Spack will optimize for the best match it can find instead
```

haampie commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -0,0 +1,708 @@
+In general, such a configuration would likely behave unexpectedly at runtime, and Spack enforces this to ensure a consistent runtime environment.
+
+The point of specs is to abstract this full DAG from Spack users.
+A user that does not care about the DAG at all, can refer to mpileaks by simply writing:
```
```suggestion
A user who does not care about the DAG at all, can refer to ``mpileaks`` by simply writing:
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-   ==> Compilers are defined in the following files:
-       /home/user/.spack/packages.yaml
+* Install further prerequisites, see :ref:`verify-spack-prerequisites`, and repeat the search above.
+* Register a buildcache that provides a compiler already available as a binary
```
That's not helpful advice for "getting started", since we do not provide this conveniently.

alalazo replied:
We have the devtools buildcaches, which are available for both linux and macOS

haampie commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,223 @@
+
+Spack lets you specify compilers on the CLI with, e.g., ``%gcc`` or ``%c,cxx=clang %fortran=gcc``, and you can specify flags with ``cflags``, ``cxxflags``, and ``fflags``.
+Depending on how complex your compiler setup is, it can be cumbersome to specify all of your preferences on the CLI.
+Spack has a special type configuration called ``toolchains``, which let you encapsulate the configuration for compilers, other libraries, and flags into a single name that you can reference as though it were one option.
```
```suggestion
Spack has a special type of configuration called ``toolchains``, which let you encapsulate the configuration for compilers, other libraries, and flags into a single name that you can reference as though it were one option.
```

haampie commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,223 @@
+When used, the toolchain above enforces using ``llvm`` for the C and C++ languages, and ``gcc`` for Fortran, when these languages are needed.
+It also adds ``cflags=-O3`` unconditionally.
+
+Using *conditional dependencies*, when constraining a virtual provider in a toolchain, is the key to make the toolchain itself widely applicable.
```
This sentence is hard to grok. Maybe rephrase entirely?

> The key concept of toolchains is requiring dependencies conditionally: if a package depends on a virtual X, then it must be provided by package Y.

Or slightly better.

I generally think the point of this paragraph doesn't land because people are not aware of the problem that's solved. Maybe introduce it with a problem, and then present toolchains as the solution.

haampie commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,223 @@
+
+to install:
+
+* An ``hdf5`` compiled with ``llvm`` for the C/C++ components, but with the fortran components compiled with ``gfortran``,
```
```suggestion
* An ``hdf5`` compiled with ``llvm`` for the C/C++ components, but with its Fortran components compiled with ``gfortran``,
```

haampie commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,223 @@
+.. command-output:: spack audit -h
+
+can be used to detect a number of configuration issues. This command detects
+configuration settings that might not be strictly wrong but are not likely
```
This is a weird way to introduce the command. You should probably say it detects potential issues with configuration and packages. It's like a linter for config and package.py files.

haampie commented on lib/spack/docs/advanced_topics.rst:
```diff
@@ -0,0 +1,223 @@
+     fcntl.lockf(self._fd, op | fcntl.LOCK_NB)
+   IOError: [Errno 38] Function not implemented
+
+A nicer error message is to be determined in future versions of Spack.
```
```suggestion
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+
+1. Specifying them as externals in ``packages.yaml``, or
+2. Having them installed in the current Spack store, or
+3. Having them available as binaries in some buildcache
```
```suggestion
3. Having them available as binaries in a buildcache
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+   -- gcc ubuntu20.04-x86_64 ---------------------------------------
+    -   gcc@12.4.0
+
+Any of these compilers can be used to build Spack packages.  More on how this is done is in :ref:`sec-specs`.
```
```suggestion
Any of these compilers can be used to build Spack packages.  More details on how this is done can be found in :ref:`sec-specs`.
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+``spack compiler info``
+-----------------------
+
+If you want to see additional information of specific compilers, you can run
```
```suggestion
If you want to see additional information about specific compilers, you can run
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+
+which will open the file in :ref:`your favorite editor <controlling-the-editor>`.
+
+Each compiler has an "external" entry in the file with some ``extra_attributes``:
```
```suggestion
Each compiler has an "external" entry in the file with ``extra_attributes``:
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+Compilers Requiring Modules
+---------------------------
+
+Many installed compilers will work regardless of the environment they are called with.
```
```suggestion
Many installed compilers will work regardless of the environment from which they are called.
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+Build Your Own Compiler
+-----------------------
+
+If you are particular about which compiler/version you use, you might wish to have Spack build it for you.
```
```suggestion
If you require a specific compiler and version, you can have Spack build it for you.
```

haampie commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -0,0 +1,292 @@
+   -- gcc ubuntu20.04-x86_64 ---------------------------------------
+   [e]  gcc@10.5.0  [+]  gcc@15.1.0  [+]  gcc@14.3.0
+
+Compilers marked with an ``[e]`` are available as externals, while those marked with a ``[+]`` are installed in the local Spack's store.
```
```suggestion
Compilers marked with an ``[e]`` are system compilers (externals), and those marked with a ``[+]`` have been installed by Spack.
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-   ==> Installing re2c-1.2.1-e3x6nxtk3ahgd63ykgy44mpuva6jhtdt
-   [ ... ]
-   zlib@1.2.11%gcc@10.1.0+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
+   The very first command run with Spack will take a while to finish, as Spack has to build a few caches to speed-up further command execution.
```
```suggestion
   The very first command run with Spack will take a while to finish, as Spack has to build a few caches to speed up subsequent command execution.
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-   [ ... ]
-   zlib@1.2.11%gcc@10.1.0+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
+   The very first command run with Spack will take a while to finish, as Spack has to build a few caches to speed-up further command execution.
+   This will be just a one-off slowdown though, and subsequent command execution is much faster.
```
```suggestion
```

That's already clear, don't spend too many words on it

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
 
-but you want to use it with Spack, you can simply run ``spack compiler find`` with the
-path to where the compiler is installed.  For example:
+If no compiler was found, you need either to:
```
```suggestion
If no compilers were found, you need either to:
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
 
-If you want to see additional information of specific compilers, you can run
-``spack compiler info``:
+Once you installed ``tcl``, you can immediately use it starting the ``tcsh`` with its absolute path:
```
```suggestion
Once you have installed ``tcl``, you can immediately use it by starting the ``tclsh`` with its absolute path:
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-.. code-block:: console
-
-   $ spack install gcc@14+binutils
+This works because of how Spack builds packages, setting the ``RPATH`` of their dependencies, but using such a long absolute path is not the most ergonomic way to refer to an executable.
```
```suggestion
This works, but using such a long absolute path is not the most convenient way to run an executable.
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
+This works because of how Spack builds packages, setting the ``RPATH`` of their dependencies, but using such a long absolute path is not the most ergonomic way to refer to an executable.
 
-Once the compiler is installed, you can start using it without additional configuration:
+The simplest way to have ``tcsh`` directory added to your ``PATH`` is to:
```
```suggestion
The simplest way to make ``tclsh`` available on the command line is:
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-^^^^^^^^^^^^^^^^^^^^^
-
-You can verify that your compilers are configured properly by installing a simple package. For example:
+Now the environment of the current shell has been modified, and you can use:
```
```suggestion
The environment of the current shell has now been modified, and you can run:
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-
-This problem is related to OpenSSL, and in some cases might be solved
-by installing a new version of ``git`` and ``openssl``:
+This section just helped you getting Spack installed, and running, quickly.
```
```suggestion
This section helped you get Spack installed and running quickly.
```

haampie commented on lib/spack/docs/getting_started.rst:
```diff
@@ -37,1396 +37,177 @@ This will create a directory called ``spack``. Once you have cloned Spack, we re
-This problem is related to OpenSSL, and in some cases might be solved
-by installing a new version of ``git`` and ``openssl``:
+This section just helped you getting Spack installed, and running, quickly.
+There are further resources in the documentation, that cover both basic and advanced topics more in details:
```
```suggestion
There are further resources in the documentation that cover both basic and advanced topics in more detail:
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+--------------------------
+
+To install software with Spack, you need to know what software is available.
+You can search the available packages at the `packages.spack.io <https://packages.spack.io>`_ website, or using the ``spack list`` command.
```
```suggestion
You can search for available packages on the `packages.spack.io <https://packages.spack.io>`_ website or by using the ``spack list`` command.
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+Packages are listed by name in alphabetical order.
+A pattern can be used to narrow the list, and the following rules apply:
+
+* A pattern to match with no wildcards, ``*`` or ``?``, will be treated as it started and ended with ``*``
```
```suggestion
* A pattern with no wildcards (``*`` or ``?``) will be treated as if it started and ended with ``*``
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+
+A few options are also provided for more specific searches.
+For instance, it is possible to search the description of packages for a match.
+A way to list all the package whose names or description contain the word ``quantum`` is the following:
```
```suggestion
A way to list all the packages whose names or descriptions contain the word ``quantum`` is the following:
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+``spack info``
+^^^^^^^^^^^^^^
+
+To get more information on a particular package from `spack list`, use
```
```suggestion
To get more information about a particular package from `spack list`, use
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+Most of the information is self-explanatory.  The *safe versions* are
+versions that Spack knows the checksum for, and it will use the
+checksum to verify that these versions download without errors or
+malware.
```
```suggestion
Most of the information is self-explanatory.
The *safe versions* are versions for which Spack knows the checksum.
Spack uses this checksum to verify that the versions are downloaded without errors or malicious changes.
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+   $ spack install mpich@3.0.4
+
+Any number of versions of the same package can be installed at once
+without interfering with each other. This is good for multi-user
```
```suggestion
without interfering with each other. This is useful for multi-user
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+
+Reusing installations in this mode is "accidental" and happening only if
+there's a match between existing installations and what Spack would have installed
+anyhow.
```
```suggestion
anyway.
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+
+   $ spack install --fresh mpich
+
+Reusing installations in this mode is "accidental" and happening only if
```
```suggestion
Reusing installations in this mode is "accidental" and happens only if
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+``spack uninstall``
+^^^^^^^^^^^^^^^^^^^
+
+To uninstall a package, type ``spack uninstall <package>``. This will ask
```
```suggestion
To uninstall a package, run ``spack uninstall <package>``. This will ask
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+^^^^^^^^^^^^^^^^^^
+
+When Spack builds software from sources, it often installs tools that are needed
+just to build or test other software. These are not necessary at runtime.
```
```suggestion
only to build or test other software. These are not necessary at runtime.
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+In the example above, Spack went through all the packages in the package database
+and removed everything that is not either:
+
+1. A package installed upon explicit request of the user
```
```suggestion
1. A package installed by explicit request of the user
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+and removed everything that is not either:
+
+1. A package installed upon explicit request of the user
+2. A ``link`` or ``run`` dependency, even transitive, of one of the packages at point 1.
```
```suggestion
In the example above, ``spack gc`` scans the package database.
It keeps only the packages that were explicitly installed by a user, along with their required ``link`` and ``run`` dependencies (including transitive dependencies).
All other packages, such as build-only dependencies or orphaned packages, are identified as "garbage" and removed.
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+
+
+Let's now say that we want to uninstall zlib. We run the command and hit a problem
+quickly because we have two!
```
```suggestion
Let's say we want to uninstall ``zlib``.
We run the command and quickly encounter a problem because two versions are installed.
```

haampie commented on lib/spack/docs/package_fundamentals.rst:
```diff
@@ -0,0 +1,1038 @@
+^^^^^^^^^^^^^^^^^^^^^^^
+
+If you have :ref:`shell support <packaging-shell-support>` enabled you can use the
+``spack load`` command to quickly get a package on your ``PATH``.
```
```suggestion
``spack load`` command to quickly add a package to your ``PATH``.
```

alalazo committed:
configuring compilers: apply suggested changes

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
getting_started.rst: apply suggested changes

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
package_fundamentals.rst: apply suggested changes

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
spec_syntax.rst: apply suggested changes

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
advanced_topics.rst: apply some of the suggested changes

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
spec_syntax.rst: implement requested changes

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
advanced_topics.rst: reword toolchains

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

tgamblin commented:
Now conflicted

alalazo committed:
Fix missing reference

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>
